include/nbl/builtin/hlsl/scan/virtual_workgroup.hlsl

```diff
@@ -2,91 +2,85 @@
 #define _NBL_HLSL_SCAN_VIRTUAL_WORKGROUP_INCLUDED_
 
 // TODO (PentaKon): Decide if these are needed once we have a clearer picture of the refactor
-#include "nbl/builtin/hlsl/limits/numeric.hlsl"
-#include "nbl/builtin/hlsl/math/typeless_arithmetic.hlsl"
+#include "nbl/builtin/hlsl/functional.hlsl"
 #include "nbl/builtin/hlsl/workgroup/arithmetic.hlsl" // This is where all the nbl_glsl_workgroupOPs are defined
 #include "nbl/builtin/hlsl/scan/declarations.hlsl"
+#include "nbl/builtin/hlsl/scan/scheduler.hlsl"
 
-#include "nbl/builtin/hlsl/binops.hlsl"
-
-#if 0
 namespace nbl
 {
 namespace hlsl
 {
 namespace scan
 {
-	template<class Binop, class Storage_t>
-	void virtualWorkgroup(in uint treeLevel, in uint localWorkgroupIndex)
-	{
-		const Parameters_t params = getParameters();
-		const uint levelInvocationIndex = localWorkgroupIndex * _NBL_HLSL_WORKGROUP_SIZE_ + SubgroupContiguousIndex();
-		const bool lastInvocationInGroup = SubgroupContiguousIndex() == (_NBL_HLSL_WORKGROUP_SIZE_ - 1);
+    template<class Binop, typename Storage_t, bool isScan, bool isExclusive, uint16_t WorkgroupSize, class Accessor, class device_capabilities=void>
+    void virtualWorkgroup(NBL_CONST_REF_ARG(uint32_t) treeLevel, NBL_CONST_REF_ARG(uint32_t) levelWorkgroupIndex, NBL_REF_ARG(Accessor) accessor)
+    {
+        const Parameters_t params = getParameters();
+        const uint32_t levelInvocationIndex = levelWorkgroupIndex * glsl::gl_WorkGroupSize().x + workgroup::SubgroupContiguousIndex();
+        const uint32_t lastLevel = params.topLevel << 1u;
+        
+        // pseudoLevel is the level index going up until toplevel then back down
+        const uint32_t pseudoLevel = treeLevel>params.topLevel ? (lastLevel-treeLevel):treeLevel;
+        const bool inRange = levelInvocationIndex <= params.lastElement[pseudoLevel];
 
-		const uint lastLevel = params.topLevel << 1u;
-		const uint pseudoLevel = levelInvocationIndex <= params.lastElement[pseudoLevel];
+        // REVIEW: Right now in order to support REDUCE operation we need to set the max treeLevel == topLevel
+        // so that it exits after reaching the top?
+        
+        // Seems that even though it's a mem barrier is must NOT diverge!
+        // This was called inside getData but due to inRange it can possibly diverge 
+        // so it must be called outside! Not yet sure why a mem barrier must be uniformly called.
+        glsl::memoryBarrierBuffer(); // scanScratchBuf can't be declared as coherent due to VMM(?)
```

Comment:
calls to pure-memory barriers (OpMemoryBarrier) can in-fact diverge, its just OpControlBarrier that needs to not diverge (same dynamic instance)
https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpControlBarrier

also you want the accessor to define the memory barrier to use